### ISSUE TITLE
docker.running ports/port_bindings not working

### ISSUE THREAD

**Issue by jasonm-makerbot** [saltstack/salt#21572]
The following is not working in the dockerio state

```
name:
  docker.running:
    - ports:
        "80/tcp":
            HostIp: "0.0.0.0"
            HostPort: "80"
```

as per docker inspect

```
    "Ports": {
        "80/tcp": null
    }
```

The ports in the container aren't being mapped back to the host.

*\* side note. The documentation shows the option name is port_bindings, it should just be "port". "port_bindings" is the name of the function in the module, seems to be an honest mistake.

**Comment by jfindlay** [comment]
@jasonm-makerbot, what is the output of `salt --versions-report`?  Thanks.

**Comment by jasonm-makerbot** [comment]
@jfindlay, We have daily builds of 2015.2
```
salt --versions-report
           Salt: 2015.2.0
         Python: 2.7.6 (default, Mar 22 2014, 22:59:56)
         Jinja2: 2.7.2
       M2Crypto: 0.21.1
 msgpack-python: 0.3.0
   msgpack-pure: Not Installed
       pycrypto: 2.6.1
        libnacl: 1.4.0
         PyYAML: 3.10
          ioflo: Not Installed
          PyZMQ: 14.0.1
           RAET: Not Installed
            ZMQ: 4.0.4
           Mako: Not Installed
```

**Comment by jasonm-makerbot** [comment]
# more info:

Salt info (*Todays build of 2015.2 at commit* c72fb4cab71f6ed6255cf8f19d92d552102d6951):
```
salt --versions-report
           Salt: 2015.2.0
         Python: 2.7.6 (default, Mar 22 2014, 22:59:56)
         Jinja2: 2.7.2
       M2Crypto: 0.21.1
 msgpack-python: 0.3.0
   msgpack-pure: Not Installed
       pycrypto: 2.6.1
        libnacl: 1.4.0
         PyYAML: 3.10
          ioflo: Not Installed
          PyZMQ: 14.0.1
           RAET: Not Installed
            ZMQ: 4.0.4
           Mako: Not Installed
```

The state:
```
docker_running:
  docker.running:
    - image: company/docker_container
    - container: docker_container
    - ports:
        "80/tcp":
            HostIp: "0.0.0.0"
            HostPort: "80"
```

The salt-master output:
```
----------
          ID: docker_running
    Function: docker.running
      Result: True
     Comment: Container 'docker_container' started.
     Started: 16:59:34.016413
    Duration: 10405.838 ms
     Changes:
              ----------
              docker_container:
                  True
----------
```

The docker inspect (partial):
```
    "NetworkSettings": {
        "Bridge": "docker0",
        "Gateway": "172.17.42.1",
        "GlobalIPv6Address": "",
        "GlobalIPv6PrefixLen": 0,
        "IPAddress": "172.17.0.34",
        "IPPrefixLen": 16,
        "IPv6Gateway": "",
        "LinkLocalIPv6Address": "fe80::42:acff:fe11:22",
        "LinkLocalIPv6PrefixLen": 64,
        "MacAddress": "02:42:ac:11:00:22",
        "PortMapping": null,
        "Ports": {
            "80/tcp": null
        }
    },
```

**Comment by jfindlay** [comment]
@jasonm-makerbot, duh, I should have remembered that you're running HEAD of 2015.2 from your other issue.  Does the syntax in [this comment](https://github.com/saltstack/salt/pull/19837#issuecomment-71087491) work (`HostIp: ''` rather than `HostIp: '0.0.0.0'`)?

**Comment by jasonm-makerbot** [comment]
@jfindlay  that doesn't help. I still get the same issue. But should that even matter if `HostIp: ''` vs `HostIp: '0.0.0.0'` ?

from `salt/states/dockerio.py`:
```
1123                 if isinstance(port[container_port], dict):
1124                     host_port = port[container_port]['HostPort']
1125                     host_ip = port[container_port].get('HostIp', '0.0.0.0')
1126                 else:
1127                     host_port = str(port[container_port])
1128                     host_ip = '0.0.0.0'
1129                 bindports[container_port] = {
1130                     'HostPort': host_port,
1131                     'HostIp': host_ip
1132                 }
```

**Comment by jfindlay** [comment]
I wonder if the problem has to do with the `port_bindings` vs `ports`.  `port_bindings` is used only inside the `running` function.  Everywhere else `ports` is used.

**Comment by jasonm-makerbot** [comment]
I think this may be related to this issue with docker-py https://github.com/docker/docker-py/issues/495

**Comment by jasonm-makerbot** [comment]
The docs here say to use `docker-py == 0.5.0`: http://docs.saltstack.com/en/latest/ref/modules/all/salt.modules.dockerio.html

`docker-py` version `0.5.0` is not in pip: https://pypi.python.org/pypi/docker-py
Earliest version of `docker-py` in it's github repo is `0.6.0` https://github.com/docker/docker-py/

Current stable version of `docker-py` is `1.1.0`

**Comment by jfindlay** [comment]
@jasonm-makerbot, we should definitely update the documentation to recommend the latest stable docker.  I don't know why the documentation is getting picky about which docker version to use.  I'm guessing that at the time there might have been some issues with other versions, so the documentation recommended a specific version that was known to work, but considering how old this is now, it's getting to be a liability.  We should probably change the documentation to recommend the latest stable.

**Comment by jasonm-makerbot** [comment]
@jfindlay I downgraded back down to `0.5.0` of `docker-py` and this appears to be working. Which is good. However, now I am afraid to ever update docker. I don't know how you want to classify this ticket now knowing this. Managing docker containers was my key reason of moving to salt since chef and puppet didn't have such features at the time, I doubt I am the only person to have such reasons so I assume possibly many others have traveled down the same path I have, but have not vocalized the issues they ran into as I have.

**Comment by jfindlay** [comment]
@jasonm-makerbot, would you mind opening a new issue specifically about the docker version support?  Thanks.

**Comment by jasonm-makerbot** [comment]
@jfindlay  https://github.com/saltstack/salt/issues/21713

**Comment by jfindlay** [comment]
thanks

**Comment by jasonm-makerbot** [comment]
@jfindlay Just want to say thank you for being on top of this issue. If you ever come east to NYC I'll grab you beer (or any beverage of your choosing as I would like to be politically correct and don't want to infer my personal morals on alcoholic beverage intake).

**Comment by jfindlay** [comment]
@jasonm-makerbot, awesome, thanks.  Docker integration is important and we need to fix this.  When you come to SLC next, I'll have @whiteinge and @cachedout and @rallytime tell you where the best local brew is to be had. :-)

**Comment by Darnok99** [comment]
try changing 
"80/tcp": 
to 
- "80/tcp":

to indicate it's a list. In the code there is a is list check

**Comment by jasonm-makerbot** [comment]
@Darnok99 I have tried this. It did not work for me.

**Comment by Darnok99** [comment]
Well in the code indeed there is a dict support, but while something like:
```
- ports:
      - '12000/tcp':
          HostIp: {{grains['ip4_interfaces']['eth0'][0]}}
          HostPort: {{settings['host-port']}}



```
salt --versions-report
           Salt: 2015.2.0rc1
         Python: 2.7.5 (default, Jun 17 2014, 18:11:42)
         Jinja2: 2.7.2
       M2Crypto: 0.21.1
 msgpack-python: 0.4.4
   msgpack-pure: Not Installed
       pycrypto: 2.6.1
        libnacl: Not Installed
         PyYAML: 3.10
          ioflo: Not Installed
          PyZMQ: 14.3.1
           RAET: Not Installed
            ZMQ: 3.2.5
           Mako: Not Installed

and docker-py 1.1.0

is working on my machine that : 

```
- ports:
      '12000/tcp':
          HostIp: {{grains['ip4_interfaces']['eth0'][0]}}
          HostPort: {{settings['host-port']}}


is not:)

**Comment by alexanderilyin** [comment]
Works in this way:

```
teamcity-server:
  docker.running:
    - container: teamcity-server-container
    - image: alexanderilyin/docker-teamcity-server
    - ports:
      - 8111: 8111
    - volumes:
      - /root/.BuildServer: /root/.BuildServer
    - watch:
      - docker: teamcity-server-container
```

with 

```
[vagrant@localhost ~]$ salt-call --version
salt-call 2015.2.0rc2 (Lithium)
```
```
[vagrant@localhost ~]$ docker -v
Docker version 1.5.0, build a8a31ef/1.5.0
```
```
[vagrant@localhost ~]$ pip show docker-py
---
Name: docker-py
Version: 1.1.0
Location: /usr/lib/python2.6/site-packages
Requires: requests, six, websocket-client
```

Doesn't work with 

```
salt-call --versions
           Salt: 2014.7.5
         Python: 2.6.6 (r266:84292, Jan 22 2014, 09:42:36)
         Jinja2: 2.2.1
       M2Crypto: 0.20.2
 msgpack-python: 0.4.6
   msgpack-pure: Not Installed
       pycrypto: 2.0.1
        libnacl: Not Installed
         PyYAML: 3.10
          ioflo: Not Installed
          PyZMQ: 14.3.1
           RAET: Not Installed
            ZMQ: 3.2.5
           Mako: Not Installed
```
```
Docker version 1.5.0, build a8a31ef/1.5.0
```
```
python-docker-py.x86_64 0.7.0-1.el6
````

**Comment by lnxnubie** [comment]
This note 
> Note: Unfortunately the documentation on the docker states is still pretty rough. Pay close attention when you are defining port bindings, they must be double indented to work properly!
from http://bitjudo.com/blog/2014/05/13/partial-continuous-deployment-with-docker-and-saltstack/
worked for me.

**Comment by ichasepucks** [comment]
@jasonm-makerbot Does your Dockerfile explicitly expose ports?  I have found that unless the ports are exposed, salt will not map them.  I have resorted to cmd.run for almost all of my docker states due to this.  Seems to not support the implicit expose as described here: http://stackoverflow.com/questions/22111060/difference-between-expose-and-publish-in-docker.  If this is not your issue, I was going to open up a new issue for this specifically.

**Comment by dimfeld** [comment]
@ichasepucks Ah, that explains the behavior I've just been seeing as well, in which the `docker run` command was able to add exposed ports to the `library/logstash` image, but it didn't work with Salt's `docker.running`.

Looking at the `docker inspect` output In my case, `NetworkSettings.Ports` and `HostConfig.PortBindings` include the port mappings, but the `Config.ExposedPorts` setting is null in the version created by salt, while in the `docker run` version it's filled in with the below:
```
 "ExposedPorts": {
     "5000/tcp": {},
     "5000/udp": {},
     "5004/udp": {}
 },```

**Comment by vitalyisaev2** [comment]
@alexanderilyin, thanks, nice workaround

**Comment by rodriguezsergio** [comment]
@@alexanderilyin got a work-around for UDP ? :stuck_out_tongue_closed_eyes:

**Comment by Deshke** [comment]
Aloha... this bug cost me 3 days to figure it out

docker.running is only using the docker option -p IF there was a EXPOSE within the docker build process. If your Dockerfile dosn't include a EXPOSE docker.running is NOT publishing the ports REGARDLESS of the ports config (i tried every combination)

EXPOSE = http://docs.docker.com/engine/reference/builder/#expose 
> The EXPOSE instruction informs Docker that the container listens on the specified network ports at runtime. EXPOSE does not make the ports of the container accessible to the host. To do that, you must use either the -p flag to publish a range of ports or the -P flag to publish all of the exposed ports. You can expose one port number and publish it externally under another number.

> To set up port redirection on the host system, see using the -P flag. The Docker network feature supports creating networks without the need to expose ports within the network, for detailed information see the overview of this feature).

Could someone PLEASE take a closer look to this one?  So "ports:" is using the docker  -P flag?  


```yml
preview-container:
  docker.running:
    - container: videoengine
    - image: tld:5043/repo/image:11c1913-1510290916
    - mem_limit: !!null
    - restart_policy:     
      MaximumRetryCount: 3,
      Name: always
    - ports:
      - "80/tcp":
            HostIp: ""
            HostPort: "80"
      - "1935/tcp":
            HostIp: ""
            HostPort: "1935"
      - "443/tcp":
            HostIp: ""
            HostPort: "443"
    - publish_all_ports: true
```
Minion:
```bash
 salt-call --versions
                  Salt: 2015.5.3
                Python: 2.7.6 (default, Jun 22 2015, 17:58:13)
                Jinja2: 2.7.2
              M2Crypto: 0.21.1
        msgpack-python: 0.3.0
          msgpack-pure: Not Installed
              pycrypto: 2.6.1
               libnacl: Not Installed
                PyYAML: 3.10
                 ioflo: Not Installed
                 PyZMQ: 14.0.1
                  RAET: Not Installed
                   ZMQ: 4.0.4
                  Mako: Not Installed
               Tornado: Not Installed
 Debian source package: 2015.5.3+ds-1trusty1

pip list
argparse (1.2.1)
backports.ssl-match-hostname (3.4.0.2)
chardet (2.0.1)
colorama (0.2.5)
croniter (0.3.4)
docker-py (1.6.0)
html5lib (0.999)
iotop (0.6)
Jinja2 (2.7.2)
M2Crypto (0.21.1)
MarkupSafe (0.18)
msgpack-python (0.3.0)
pip (1.5.4)
pycrypto (2.6.1)
pycurl (7.19.3)
pyinotify (0.9.4)
python-apt (0.9.3.5ubuntu1)
python-dateutil (1.5)
python-debian (0.1.21-nmu2ubuntu2)
PyYAML (3.10)
pyzmq (14.0.1)
requests (2.8.1)
salt (2015.5.3)
setuptools (3.3)
six (1.10.0)
ssh-import-id (3.21)
urllib3 (1.7.1)
websocket-client (0.34.0)
wheel (0.24.0)
wsgiref (0.1.2)

```

**Comment by vitalyisaev2** [comment]
@Deshke I can confirm that port mapping is still very tricky and I would suggest you to use workaround of @alexanderilyin  (see above). Probably like this:
```
...
- ports:
      - 80:80
      - 433:433
      - 1935:1935
```

**Comment by Deshke** [comment]
the workaround didn't work - i tried that... on docker-py (1.6.0)

```
- ports:
      - 80:80
      - 433:433
      - 1935:1935

```
throws a syntax err it should be

```yaml
- ports:
      - 80: 80
      - 433: 433
      - 1935: 1935

```

**Comment by vitalyisaev2** [comment]
We use ```docker``` 1.6 with ```docker-py``` 1.2.3. We store port mappings in Pillars, not in Salt States. Consider this:

```
➜  salt-master git:(master) cat pillar/restapi.sls

{% set restapi = pillar["restapi"] %}

(...)

restapi:
  root: restapi
  node: restapi
  repo: mailpaas
  image: restapi-validators
  ports:
    - 8080: 8080
    - 20003: 20003
  locators:
    - '{{ ag.server.host }}:{{ ag.server.port }}'
  service:
    name: restapi
    tag: {{ prefix }}
  server:
    host: {{ cfg[grains.node_use] }}
    port: 8080
  handystats:
    port: 20003
```

Than we use the mappings in States:
```
➜  salt-master git:(master) cat salt/mailpaas/restapi.sls

(...)

{{ restapi_node }}:
  docker.running:
    - container: {{ restapi_node }}
    - image: {{ restapi_image }}:{{ label }}
    - volumes:
      - /dev/log : /dev/log
      - {{ restapi_external_log_dir }} : /var/log
      - {{ restapi_external_conf_dir }} : {{ restapi_internal_conf_dir }}
    - ports: {{ restapi['ports'] }}
{% if grains.orchestration == "consul" %}
    - environment:
      - SERVICE_{{ restapi.server.port }}_NAME: {{ restapi["service"]["name"] }}
      - SERVICE_{{ restapi.handystats.port }}_NAME: {{ restapi["service"]["name"] }}-stats
      - SERVICE_TAGS: {{ restapi["service"]["tag"] }}
{% endif %}
    - command: -c {{ restapi_internal_conf_dir }}/restapi.json
    - require:
      - pip: docker-py
      - docker: {{ restapi_node }}_image
```

**Comment by Deshke** [comment]
state or pillar shouldn't make a difference here :) btw. using docker 1.9.1 with the latest docker-py (1.6) as mentioned in https://github.com/saltstack/salt/issues/23638#issuecomment-161931834

**Comment by vitalyisaev2** [comment]
@Deshke , I've just only provided a working example in order to give you opportunity to check indentation in your ```ports``` variable and so on. Hope it helps
